designs/statefulset-disruption.md

```diff
@@ -0,0 +1,891 @@
+
+As of EBS CSI Driver `v1.31.0`, the EBS CSI Node Pod tolerates all taints be default, therefore we will focus on this second type of race for the following diagram:
+
+![6min](https://github.com/AndrewSirenko/karpenter-provider-aws/assets/68304519/0a0fbe16-fe81-428d-85fb-0716c733c9c0)
+
+Karpenter's terminator cannot drain pods that tolerate its `Disrupting` taint. Therefore, once it drains the drainable pods, it calls EC2 TerminateInstance on a node. 
+
+However, the shutdown manager does not wait for all volumes to be unmounted, just pod termination. This leads to a race condition where the CSI Driver Node Pod is killed before all unmounts are completed. See @msau42's diagram:
+
+![kubelet_shutdown_race](https://github.com/AndrewSirenko/karpenter-provider-aws/assets/68304519/9ccb6c94-ef02-40af-84bf-9471827f3303)
+
+Today, the EBS CSI Driver attempts to work around these races by utilizing a [PreStop hook](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/docs/faq.md#what-steps-can-be-taken-to-mitigate-this-issue) that tries to keep the Node Service alive for an additional [terminationGracePeriodSeconds](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/73b43995a70b8fea51655577e4b475613ab400f1/charts/aws-ebs-csi-driver/values.yaml#L373) until all volumes are unmounted. We will explore the shortcomings of this solution later in [problem A alternative solutions](#alternatives-considered).  
+
+*Note: In addition to the Kubelet race, this delay can happen if stateful pods out-live the CSI Node Pod. E.g. Operator has a statefulset that tolerates all taints and has a longer terminationGracePeriod than EBS CSI Driver.*
+
+### Solutions:
+
+We should:
+- A1: Fix Kubelet race condition upstream for future Kubernetes versions
+- A2: Have Karpenter taint terminated nodes as `out-of-service` before removing its finalizer. 
+
+#### A1: Fix race at Kubelet level
+
+The Kubelet Shutdown Manager should not kill CSI Driver Pods before volumes are unmounted. This change must be made at `kubernetes/kubernetes` level. 
+
+Because this solution does not rely on changes in Karpenter, please see [Active PR #125070](https://github.com/kubernetes/kubernetes/pull/125070) for more information on this solution. 
+
+**Pros:** 
+
+- Other cluster autoscalers will not face this race condition. 
+- Reduces pod migration times
+- Reduces risk of data corruption because relevant CSI Drivers will perform unmount operations required  
+
+**Cons:**
+
+- Unavailable until merged in a version of Kubernetes (Likely Kubernetes `v1.31`) (Possibly able to be backported)
+- If Kubelet Shutdown Manager is not able to run 
```

Comment:
Do we know how often this happens?

Reply:
Typically depends on the fix. We will know once relevant SIG reviewer leaves review on PR.

Reply:
i'm not sure I understand. Sounds like you're saying that the amount of times that it won't be able to run will be different after the upstream PR is merged? what is the current state like? Mostly just trying to understand what causes this to happen that Karpenter can be aware of.

Reply:
Was looking at wrong line. Shutdown manager will **always** run IF graceful shutdown is enabled properly on node. Will run until gracefulShutdown time is up. If volumes are NOT unmounted by CSI Node pod by then (gracefulShutdown period being up) for some reason, then kubernetes cannot confirm volume was not unmounted, and volume will still have delay.